src/pyhf/workspace.py

```diff
@@ -16,11 +18,263 @@
+
+    """
+    if left_version != right_version:
+        raise exceptions.InvalidWorkspaceOperation(
+            f"Workspaces of different versions cannot be combined: {left_version} != {right_version}"
+        )
+    return left_version
+
+
+def _join_channels(join, left_channels, right_channels):
+    """
+    Join two workspace channel specifications.
+
+    Raises:
+      ~pyhf.exceptions.InvalidWorkspaceOperation: Channel specifications are incompatible.
+
+    Args:
+        join (str): The join operation to apply. See ~pyhf.workspace.Workspace for valid join operations.
+        left_channels (list): The left channel specification.
+        right_channels (list): The right channel specification.
+
+    Returns:
+        :obj:`list`: A joined list of channels. Each channel follows the :obj:`defs.json#channel` `schema <https://scikit-hep.org/pyhf/likelihood.html#channel>`__
+
+    """
+
+    joined_channels = _join_items(join, left_channels, right_channels)
+    if join == 'none':
+        common_channels = set(c['name'] for c in left_channels).intersection(
+            c['name'] for c in right_channels
+        )
+        if common_channels:
+            raise exceptions.InvalidWorkspaceOperation(
+                f"Workspaces cannot have any channels in common with the same name: {common_channels}. You can also try a different join operation: {Workspace.valid_joins}."
+            )
+
+    elif join == 'outer':
+        counted_channels = collections.Counter(
+            channel['name'] for channel in joined_channels
+        )
+        incompatible_channels = [
+            channel for channel, count in counted_channels.items() if count > 1
+        ]
+        if incompatible_channels:
+            raise exceptions.InvalidWorkspaceOperation(
+                f"Workspaces cannot have channels in common with incompatible structure: {incompatible_channels}. You can also try a different join operation: {Workspace.valid_joins}."
+            )
+    return joined_channels
+
+
+def _join_observations(join, left_observations, right_observations):
+    """
+    Join two workspace observation specifications.
+
+    Raises:
+      ~pyhf.exceptions.InvalidWorkspaceOperation: Observation specifications are incompatible.
+
+    Args:
+        join (str): The join operation to apply. See ~pyhf.workspace.Workspace for valid join operations.
+        left_observations (list): The left observation specification.
+        right_observations (list): The right observation specification.
+
+    Returns:
+        :obj:`list`: A joined list of observations. Each observation follows the :obj:`defs.json#observation` `schema <https://scikit-hep.org/pyhf/likelihood.html#observations>`__
+
+    """
+    joined_observations = _join_items(join, left_observations, right_observations)
+    if join == 'none':
+        common_observations = set(c['name'] for c in left_observations).intersection(
+            c['name'] for c in right_observations
```

Comment:
I might be tempted to use `obs` instead of `c` here for clarity, but this is more personal preference.